text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+implement `Clone` and `Copy`.
+
+## Restrictions in compound types
+[restrictions-on-compound-types]: #restrictions-on-compound-types
+
+`Pointee` types cannot be used in non-`#[repr(transparent)]` compound types
+as the alignment of these types would need to be known in order to calculate field
+offsets. `const Sized` types can be used in compound types with no restrictions.
+`Sized`, `const ValueSized` and `ValueSized` types can be used in compound types, but
+only as the last element.
+
+## Compiler performance implications
+[compiler-performance-implications]: #compiler-performance-implications
+
+There is a potential performance impact within the trait system to adding
+supertraits to `Sized`, as implementation of these supertraits will need to be
+proven whenever a `Sized` obligation is being proven (and this happens very
+frequently, being a default bound). It may be necessary to implement an
+optimisation whereby `Sized`'s supertraits are assumed to be implemented and
+checking them is skipped - this should be sound as all of these traits are
+implemented by the compiler and therefore this property can be guaranteed.
+
+## Forward compatibility with supertraits implying default bound removal
+[forward-compatibility-with-supertraits-default-bound]: #forward-compatibility-with-supertraits-implying-default-bound-removal
+
+Traits which are a supertrait of any of the proposed traits will not
+automatically imply the proposed trait in any bounds where the trait is
+used, e.g.
+
+```rust
+trait NewTrait: ValueSized {}
+
+struct NewRc<T: NewTrait> {} // equiv to `T: NewTrait + Sized` as today
+```
+
+If the user wanted `T: ValueSized` then it would need to be written explicitly.
+This is forward compatible with trait bounds which have sizedness supertraits
+implying the removal of the default `const Sized` bound.
+
+## Ecosystem churn
+[ecosystem-churn]: #ecosystem-churn
+
+It is not expected that this RFC's additions would result in much churn within
+the ecosystem. Almost all of the necessary changes would happen automatically
+during edition migration.
+
+All bounds in the standard library should be re-evaluated during the
+implementation of this RFC, but bounds in third-party crates need not be.
+
+As runtime-sized types will primarily be used for localised performance optimisation,
+and `Pointee` types will primarily be used for localised FFI, neither is expected
+to be so pervasive throughout Rust codebases to the extent that all existing
+`const Sized`, `~const ValueSized` or `ValueSized` bounds (after edition migration) would
+need to be immediately reconsidered in light of their addition, even if in many cases
+these could be relaxed.
+
+If a user of a runtime-sized type or a `Pointee` type did encounter a bound that
+needed to be relaxed, this could be changed in a patch to the relevant crate without
+breaking backwards compatibility as-and-when such cases are encountered.
```

Comment:
I think this is an overly rosy outlook. It may be very annoying for types that aren't `const ValueSized` to be locked out of implementing traits from other crates as it will make it much harder for them to be used like normal types, also we'd need to teach crate authors to write new code as permissively as possible. 
Additionally it's not always going to be possible to relax these bounds without a breaking change, my personally scariest trait is [serde's `Serializer` trait](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_some) as it's impossible to relax that bound as serializers might rely on it but it prevents these new types from being first class members of the swede ecosystem.

Reply:
Could you elaborate on this point a bit and spell it out further?  I feel like there may be something important here to consider, and am interested in seeing it detailed.

Reply:
Here's a more self contained example consider a library with the following trait:

```rust

trait A {
  fn a<T: ?Sized>(&mut self, x: &T);
}

```

An impl of that trait in a different crate may call `size_of_val`,  so the crate that defines A can't backwards compatibly relax the type of T to Pointee instead of `const ValueSized`.

Reply:
A problematic case of this in std is Deref. 

Ideally, I think that the Target type should be relaxed to Pointee, but that wouldn't be backwards compatible, because existing code might depend on being able to call `size_of_val` on `&<T as Deref>::Target`.

Reply:
As mentioned [here](https://github.com/rust-lang/rfcs/pull/3729#discussion_r1851828543), I think that it is possible to relax associated types in a backwards compatible way, as long as we do it with the edition migration required for this RFC. However, I agree with your minimal example on a situation where I don't think there's any backwards compatible solution.

Reply:
This is a good catch, I didn't realise trait methods would an issue. I don't have a good answer to how to deal with those, but I've noted this in the RFC.